openpgp/packet/private_key.go

```diff
@@ -278,19 +364,60 @@ func (pk *PrivateKey) Serialize(w io.Writer) (err error) {
 		return
 	}
 
+	
+
 	optional := bytes.NewBuffer(nil)
 	if pk.Encrypted || pk.Dummy() {
-		optional.Write([]byte{uint8(pk.cipher)})
-		if err := pk.s2kParams.Serialize(optional); err != nil {
+		// [Optional] If string-to-key usage octet was 255, 254, or 253,
+		// a one-octet symmetric encryption algorithm.
+		if _, err = optional.Write([]byte{uint8(pk.cipher)}); err != nil {
+			return
+		}
+		// [Optional] If string-to-key usage octet was 253, 
+		// a one-octet AEAD algorithm.
+		if pk.s2kType == S2KAEAD {
+			if _, err = optional.Write([]byte{uint8(pk.aead)}); err != nil {
+				return
+			}
+		}
+
+		s2kBuffer := bytes.NewBuffer(nil)
+		if err := pk.s2kParams.Serialize(s2kBuffer); err != nil {
 			return err
 		}
+		// [Optional] Only for a version 6 packet, and if string-to-key 
+		// usage octet was 255, 254, or 253, an one-octet 
+		// count of the following field.
+		if pk.Version == 6 {
+			if _, err = optional.Write([]byte{uint8(s2kBuffer.Len())}); err != nil {
+				return
+			}
+		} 
+		// [Optional] If string-to-key usage octet was 255, 254, or 253, 
+		// a string-to-key (S2K) specifier. The length of the string-to-key specifier 
+		// depends on its type 
+		if _, err = io.Copy(optional, s2kBuffer); err != nil {
+			return
+		}
+
+		// IV
 		if pk.Encrypted {
-			optional.Write(pk.iv)
+			if _, err = optional.Write(pk.iv); err != nil {
+				return
+			}
+			if pk.Version == 5 && len(pk.iv) > 0 {
```

Comment:
Maybe also make this
```suggestion
			if pk.Version == 5 && pk.s2kType == S2KAEAD {
```
to make it more obvious that that's where it's needed?